README.md
# Bitcraft
> A toolkit for bit-strings and DSL for binary protocols bit-blocks.

![CI](https://github.com/alboratech/bitcraft/workflows/CI/badge.svg)
[![Coverage Status](https://img.shields.io/coveralls/alboratech/bitcraft.svg)](https://coveralls.io/github/alboratech/bitcraft)
[![Hex Version](https://img.shields.io/hexpm/v/bitcraft.svg)](https://hex.pm/packages/bitcraft)
[![Docs](https://img.shields.io/badge/docs-hexpm-blue.svg)](https://hexdocs.pm/bitcraft)
[![License](https://img.shields.io/hexpm/l/bitcraft.svg)](LICENSE)

When working with binary protocols we usually have to implement encoding and
decoding functions for the different type of messages the protocol supports.
Despite parsing binary protocols is relatively easy in Elixir/Erlang using
binary pattern-matching (and one of the greatest features in Elixir/Erlang),
it might be tedius implement X number of parsing functions to support the
protocol messages, we may ending up with a lot of similar binary matching
all over the code, which is not bad, but what if we could avoid it?
What if we had a toolkit like **Ecto** to define parseable bit-blocks,
commonly used in binary protocols? This is where **Bitcraft** comes in!

**Bitcraft** provides a DSL for defining parseable binary blocks or messages.
You just need to define the bit-block for your message, adding the segments
with their names, sizes and properties, and then **Bitcraft** generates the
encoding and decoding functions automatically.

## Installation

You need to add `bitcraft` as a dependency to your `mix.exs` file.

```elixir
def deps do
  [
    {:bitcraft, "~> 0.1.0"}
  ]
end
```

## Getting started

Let's start defining the bit-block for a simple message:

```elixir
defmodule MyBlock do
  import Bitcraft.BitBlock

  defblock "my-block" do
    segment(:header, 5, type: :binary)
    segment(:s1, 4, default: 1)
    segment(:s2, 8, default: 1, sign: :signed)
    segment(:tail, 3, type: :binary)
  end
end
```

After compile your code, you will be able to run:

```elixir
iex> data = %MyBlock{header: "begin", s1: 3, s2: -3, tail: "end"}
iex> bits = MyBlock.encode(block)
<<98, 101, 103, 105, 110, 63, 214, 86, 230, 4::size(4)>>

iex> MyBlock.decode(bits)
%MyBlock{header: "begin", leftover: "", s1: 3, s2: -3, tail: "end"}
```

> See `Bitcraft.BitBlock.defblock/3` and `Bitcraft.BitBlock.segment/3`
  for more information.

### Working with dynamic blocks

For this example let's define an IPv4 datagram, which has a dynamic part:

```elixir
defmodule IpDatagram do
  import Bitcraft.BitBlock

  defblock "IP-datagram" do
    segment(:vsn, 4)
    segment(:hlen, 4)
    segment(:srvc_type, 8)
    segment(:tot_len, 16)
    segment(:id, 16)
    segment(:flags, 3)
    segment(:frag_off, 13)
    segment(:ttl, 8)
    segment(:proto, 8)
    segment(:hdr_chksum, 16, type: :bits)
    segment(:src_ip, 32, type: :bits)
    segment(:dst_ip, 32, type: :bits)
    segment(:opts, :dynamic, type: :bits)
    segment(:data, :dynamic, type: :bits)
  end

  # Size resolver for dynamic segments invoked during the decoding
  def calc_size(%__MODULE__{hlen: hlen}, :opts, dgram_s)
      when hlen >= 5 and 4 * hlen <= dgram_s do
    opts_s = 4 * (hlen - 5)
    {opts_s * 8, dgram_s}
  end

  def calc_size(%__MODULE__{leftover: leftover}, :data, dgram_s) do
    data_s = :erlang.bit_size(leftover)
    {data_s, dgram_s}
  end
end
```

Here, the segment corresponding to the `:opts` segment has a type modifier,
specifying that `:opts` is to bind to a bitstring (or binary). All other
segments have the default type equal to unsigned integer.

An IP datagram header is of variable length. This length is measured in the
number of 32-bit words and is given in the segment corresponding to `:hlen`.
The minimum value of `:hlen` is 5. It is the segment corresponding to
`:opts` that is variable, so if `:hlen` is equal to 5, `:opts` becomes
an empty binary. Finally, the tail segment `:data` bind to bitstring.

The decoding of the datagram fails if one of the following occurs:

  * The first 4-bits segment of datagram is not equal to 4.
  * `:hlen` is less than 5.
  * The size of the datagram is less than `4*hlen`.

Since this block has dynamic segments, we can now use the other decode
arguments to resolve the size for them during the decoding process:

```elixir
IpDatagram.decode(bits, :erlang.bit_size(bits), &IpDatagram.calc_size/3)
```

Where:

  * The first argument is the input IPv4 datagram (bitstring).
  * The second argument is is the accumulator to the callback function
    (third argument), in this case is the total number of bits in the
    datagram.
  * And the third argument is the function callback or dynamic size resolver
    that will be invoked by the decoder for each dynamic segment. The callback
    functions receives the data struct with the current decoded segments, the
    segment name (to be pattern-matched and resolve its size), and the
    accumulator that can be used to pass metadata during the dynamic
    segments evaluation.

It is time to try it out! First of all, let's create a `IpDatagram` data type
with valid data:

```elixir
iex> dgram = %IpDatagram{
...>   vsn: 4,
...>   hlen: 6,
...>   srvc_type: 8,
...>   tot_len: 100,
...>   id: 1,
...>   flags: 1,
...>   frag_off: 1,
...>   ttl: 32,
...>   proto: 6,
...>   hdr_chksum: <<1, 1>>,
...>   src_ip: <<10, 10, 10, 5>>,
...>   dst_ip: <<10, 10, 10, 6>>,
...>   opts: %Bitcraft.BitBlock.DynamicSegment{
...>     value: <<10, 10, 10, 1>>,
...>     size: 32
...>   },
...>   data: %Bitcraft.BitBlock.DynamicSegment{
...>     value: "ping",
...>     size: 32
...>   }
...> }
```

As you notice, for the dynamic segments we use the data type
`Bitcraft.BitBlock.DynamicSegment` type, with the corresponding value and size
in bits. This will tell `Bitcraft` how the block should be encoded. This is the
way to set dynamic segments, the value cannot be assigned directly, it is to be
encapsulated within this data type with the value and size.

Now let's encode it:

```elixir
iex> bits = IpDatagram.encode(dgram)
<<70, 8, 0, 100, 0, 1, 32, 1, 32, 6, 1, 1, 10, 10, 10, 5, 10, 10, 10, 6, 10, 10,
  10, 1, 112, 105, 110, 103>>
```

Finally, for decoding it, we have to use the callback to resolve the dynamic
sizes, which was defined previously within the module `IpDatagram.calc_size/3`.

```elixir
iex> IpDatagram.decode(bits, :erlang.bit_size(bits), &IpDatagram.calc_size/3)
%IpDatagram{
  data: %Bitcraft.BitBlock.DynamicSegment{size: 32, value: "ping"},
  dst_ip: <<10, 10, 10, 6>>,
  flags: 1,
  frag_off: 1,
  hdr_chksum: <<1, 1>>,
  hlen: 6,
  id: 1,
  leftover: "",
  opts: %Bitcraft.BitBlock.DynamicSegment{size: 32, value: <<10, 10, 10, 1>>},
  proto: 6,
  src_ip: <<10, 10, 10, 5>>,
  srvc_type: 8,
  tot_len: 100,
  ttl: 32,
  vsn: 4
}
```

## Arrays

Sometimes we may also want to parse a segment of bits as an array, for example,
suppose we have a dynamic segment but we want to parse it as a list of integers
of 16 bits, what variates is the amount of them (the length of the array).
So, if the size of the segment is calculated as 64 bits, we expect an array
of 4 integers of 16 bits.

First of all, let us define a block with an array-type segment:

```elixir
defmodule TestBlock do
  import Bitcraft.BitBlock

  defblock "test-block" do
    segment(:a, 8)
    segment(:b, 8)
    array(:list, type: :integer, element_size: 16, sign: :signed)
  end

  # Size resolver for dynamic segments invoked during the decoding
  def calc_size(%__MODULE__{a: a, b: b}, :list, acc) do
    {a + b, acc}
  end
end
```

As you may notice, the field `:list` is defined as array-type in the form:

```elixir
array(:list, type: :integer, element_size: 16, sign: :signed)
```

The first argument is the name of the segment, then we pass the options.
The `type: :integer` defines the the type for the array elements, and
`element_size: 16` defines that each element of the array must have
16 bits size. The rest of the options are the same and apply to the
array elements.

> See `Bitcraft.BitBlock.array/2` and `Bitcraft.BitBlock.segment/3`.

Now, let's try it out:

```elixir
iex> data = %TestBlock{
...>   a: 32,
...>   b: 32,
...>   list: %Bitcraft.BitBlock.DynamicSegment{value: [1, 2, 3, 4], size: 64}
}
iex> encoded = TestBlock.encode(data)
<<32, 32, 0, 1, 0, 2, 0, 3, 0, 4>>
iex> TestBlock.decode(encoded, %{}, &TestBlock.calc_size/3)
%TestBlock{
  a: 32,
  b: 32,
  leftover: "",
  list: %Bitcraft.BitBlock.DynamicSegment{size: 64, value: [1, 2, 3, 4]}
}
```

## Contributing

Contributions to Bitcraft are very welcome and appreciated!

Use the [issue tracker](https://github.com/alboratech/bitcraft/issues) for bug
reports or feature requests. Open a
[pull request](https://github.com/alboratech/bitcraft/pulls)
when you are ready to contribute.

When submitting a pull request you should not update the
[CHANGELOG.md](CHANGELOG.md), and also make sure you test
your changes thoroughly, include unit tests
alongside new or changed code.

Before to submit a PR it is highly recommended to run `mix check` and ensure all
checks run successfully.

## Copyright and License

Copyright (c) 2020, Carlos Bolaños.

Copyright (c) 2021, Albora Technologies Ltd.

Bitcraft source code is licensed under the [MIT License](LICENSE).


CHANGELOG.md
Changelog
=========

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.1.0]

### Added

- Added DSL for defining bit blocks with their segments and provide
  encode/decode functions ot-of-box.


test/bitcraft_test.exs
defmodule BitcraftTest do
  @moduledoc false
  use ExUnit.Case
  doctest Bitcraft

  alias Bitcraft.BitBlock.Array

  describe "encode_segment/2 & decode_segment/2" do
    test "integers" do
      assert 5 |> Bitcraft.encode_segment() |> Bitcraft.decode_segment() == {5, ""}
      assert 3 |> Bitcraft.encode_segment(size: 4) |> Bitcraft.decode_segment(size: 4) == {3, ""}

      assert -3
             |> Bitcraft.encode_segment(size: 4, sign: :signed)
             |> Bitcraft.decode_segment(size: 4, sign: :signed) == {-3, ""}

      assert -3
             |> Bitcraft.encode_segment(size: 4, sign: :signed, endian: :little)
             |> Bitcraft.decode_segment(size: 4, sign: :signed, endian: :little) == {-3, ""}
    end

    test "floats" do
      assert 5.5
             |> Bitcraft.encode_segment(type: :float)
             |> Bitcraft.decode_segment(type: :float) == {5.5, ""}

      assert 3.3
             |> Bitcraft.encode_segment(size: 64, type: :float)
             |> Bitcraft.decode_segment(size: 64, type: :float) == {3.3, ""}

      assert -3.3
             |> Bitcraft.encode_segment(size: 64, type: :float, sign: :signed)
             |> Bitcraft.decode_segment(size: 64, type: :float, sign: :signed) == {-3.3, ""}

      assert -3.3
             |> Bitcraft.encode_segment(size: 64, type: :float, sign: :signed, endian: :little)
             |> Bitcraft.decode_segment(size: 64, type: :float, sign: :signed, endian: :little) ==
               {-3.3, ""}
    end

    test "binaries/bits" do
      assert "test"
             |> Bitcraft.encode_segment(type: :binary)
             |> Bitcraft.decode_segment(size: 4, type: :binary) == {"test", ""}

      assert <<1, 2, 3, 4>>
             |> Bitcraft.encode_segment(type: :bits)
             |> Bitcraft.decode_segment(size: 32, type: :bits) == {<<1, 2, 3, 4>>, ""}
    end

    test "utf8/utf16/utf32" do
      assert ?x
             |> Bitcraft.encode_segment(type: :utf8)
             |> Bitcraft.decode_segment(type: :utf8) == {"x", ""}

      assert "x"
             |> Bitcraft.encode_segment(type: :utf8)
             |> Bitcraft.decode_segment(type: :utf8) == {"x", ""}

      assert "foo"
             |> Bitcraft.encode_segment(type: :utf16)
             |> Bitcraft.decode_segment(type: :utf16) == {"foo", ""}

      assert "foo"
             |> Bitcraft.encode_segment(type: :utf32)
             |> Bitcraft.decode_segment(type: :utf32) == {"foo", ""}
    end

    test "array" do
      assert [1, 2, 3]
             |> Bitcraft.encode_segment(type: %Array{})
             |> Bitcraft.decode_segment(size: 24, type: %Array{}) == {[1, 2, 3], ""}

      assert [250, 252, 255]
             |> Bitcraft.encode_segment(type: %Array{element_size: 4})
             |> Bitcraft.decode_segment(size: 12, type: %Array{element_size: 4}) ==
               {[10, 12, 15], ""}

      assert [3.3, -7.7, 9.9]
             |> Bitcraft.encode_segment(
               type: %Array{type: :float, element_size: 64},
               sign: :signed
             )
             |> Bitcraft.decode_segment(
               size: 192,
               type: %Array{type: :float, element_size: 64},
               sign: :signed
             ) == {[3.3, -7.7, 9.9], ""}

      assert ["hello", "world"]
             |> Bitcraft.encode_segment(type: %Array{type: :binary, element_size: 5})
             |> Bitcraft.decode_segment(size: 10, type: %Array{type: :binary, element_size: 5}) ==
               {["hello", "world"], ""}

      assert [<<1, 2, 3>>, <<4, 5, 6>>]
             |> Bitcraft.encode_segment(type: %Array{type: :bits, element_size: 24})
             |> Bitcraft.decode_segment(size: 48, type: %Array{type: :bits, element_size: 24}) ==
               {[<<1, 2, 3>>, <<4, 5, 6>>], ""}
    end
  end

  describe "count_ones/1" do
    test "returns the number of 1's bits for the given integer" do
      assert Bitcraft.count_ones(1) == 1
      assert Bitcraft.count_ones(3) == 2
      assert Bitcraft.count_ones(15) == 4
    end
  end
end


test/support/test_blocks.ex
defmodule Bitcraft.TestBlocks do
  @moduledoc false

  defmodule StaticBlock do
    @moduledoc false
    import Bitcraft.BitBlock

    defblock "test-block-1" do
      segment(:header, 5, type: :binary)
      segment(:a, 4, default: 1)
      segment(:b, 8, default: 1)
      segment(:c, 16, default: 1, sign: :signed)
      segment(:tail, 3, type: :binary)
    end

    def sample do
      %__MODULE__{
        header: "begin",
        a: 3,
        b: 5,
        c: -10_000,
        tail: "end"
      }
    end
  end

  defmodule DynamicBlock do
    @moduledoc false
    import Bitcraft.BitBlock

    alias Bitcraft.BitBlock.DynamicSegment

    @type t :: %__MODULE__{}

    defblock "test-block-1", typespec: false do
      segment(:header, 4, type: :binary)
      segment(:a, 4, default: 1)
      segment(:b, 8, default: 1)
      segment(:c, 16, default: 1, sign: :signed)
      segment(:tail, 8, type: :utf8)
      segment(:d, :dynamic)
      array(:e, type: :integer, element_size: 4, sign: :signed)
      segment(:extra)
    end

    def callback(%__MODULE__{a: a, b: b}, :d, acc) do
      d_size = Bitcraft.count_ones(a * b)
      {d_size, Map.put(acc, :nd, d_size)}
    end

    def callback(%__MODULE__{}, :e, %{nd: nd} = acc) do
      {nd * 4, acc}
    end

    def sample do
      %__MODULE__{
        header: "test",
        a: 3,
        b: 5,
        c: -10_000,
        tail: ?x,
        d: %DynamicSegment{
          value: 2,
          size: 4
        },
        e: %DynamicSegment{
          value: [1, -1, 2, -2],
          size: 16
        }
      }
    end
  end

  defmodule IpDatagram do
    @moduledoc false
    import Bitcraft.BitBlock

    alias Bitcraft.BitBlock.DynamicSegment

    defblock "IP-datagram" do
      segment(:vsn, 4)
      segment(:hlen, 4)
      segment(:srvc_type, 8)
      segment(:tot_len, 16)
      segment(:id, 16)
      segment(:flags, 3)
      segment(:frag_off, 13)
      segment(:ttl, 8)
      segment(:proto, 8)
      segment(:hdr_chksum, 16, type: :bits)
      segment(:src_ip, 32, type: :bits)
      segment(:dst_ip, 32, type: :bits)
      segment(:opts, :dynamic, type: :bits)
      segment(:data, :dynamic, type: :bits)
    end

    def callback(%__MODULE__{hlen: hlen}, :opts, dgram_s)
        when hlen >= 5 and 4 * hlen <= dgram_s do
      opts_s = 4 * (hlen - 5)
      {opts_s * 8, dgram_s}
    end

    def callback(%__MODULE__{leftover: leftover}, :data, dgram_s) do
      data_s = :erlang.bit_size(leftover)
      {data_s, dgram_s}
    end

    def sample do
      %__MODULE__{
        vsn: 4,
        hlen: 6,
        srvc_type: 8,
        tot_len: 100,
        id: 1,
        flags: 1,
        frag_off: 1,
        ttl: 32,
        proto: 6,
        hdr_chksum: <<1, 1>>,
        src_ip: <<10, 10, 10, 5>>,
        dst_ip: <<10, 10, 10, 6>>,
        opts: %DynamicSegment{
          value: <<10, 10, 10, 1>>,
          size: 32
        },
        data: %DynamicSegment{
          value: "ping",
          size: 32
        }
      }
    end
  end
end


test/bitcraft/bit_block_test.exs
defmodule Bitcraft.BitBlockTest do
  @moduledoc false
  use ExUnit.Case
  doctest Bitcraft

  alias Bitcraft.BitBlock.DynamicSegment
  alias Bitcraft.TestBlocks.{DynamicBlock, IpDatagram, StaticBlock}

  describe "encode/decode" do
    test "static block works properly" do
      assert bits = StaticBlock.sample() |> StaticBlock.encode()
      assert data = StaticBlock.decode(bits)

      assert data == %StaticBlock{
               a: 3,
               b: 5,
               c: -10_000,
               header: "begin",
               leftover: "",
               tail: "end"
             }

      assert data |> StaticBlock.encode() == bits
    end

    test "dynamic block works properly" do
      assert bits = DynamicBlock.sample() |> DynamicBlock.encode()
      assert data = DynamicBlock.decode(bits, %{}, &DynamicBlock.callback/3)

      assert data == %DynamicBlock{
               header: "test",
               a: 3,
               b: 5,
               c: -10_000,
               tail: 120,
               d: %DynamicSegment{size: 4, value: 2},
               e: %DynamicSegment{size: 16, value: [1, -1, 2, -2]},
               leftover: "",
               extra: nil
             }

      assert data |> DynamicBlock.encode() == bits
    end

    test "IP datagram" do
      assert bits = IpDatagram.sample() |> IpDatagram.encode()
      assert data = IpDatagram.decode(bits, %{}, &IpDatagram.callback/3)

      assert data == %IpDatagram{
               data: %DynamicSegment{size: 32, value: "ping"},
               dst_ip: <<10, 10, 10, 6>>,
               flags: 1,
               frag_off: 1,
               hdr_chksum: <<1, 1>>,
               hlen: 6,
               id: 1,
               leftover: "",
               opts: %DynamicSegment{size: 32, value: <<10, 10, 10, 1>>},
               proto: 6,
               src_ip: <<10, 10, 10, 5>>,
               srvc_type: 8,
               tot_len: 100,
               ttl: 32,
               vsn: 4
             }

      assert data |> IpDatagram.encode() == bits
    end

    test "IP datagram without opts" do
      dgram = IpDatagram.sample()
      assert bits = IpDatagram.encode(%{dgram | hlen: 5, opts: nil})
      assert data = IpDatagram.decode(bits, :erlang.bit_size(bits), &IpDatagram.callback/3)

      assert data == %IpDatagram{
               data: %DynamicSegment{size: 32, value: "ping"},
               dst_ip: <<10, 10, 10, 6>>,
               flags: 1,
               frag_off: 1,
               hdr_chksum: <<1, 1>>,
               hlen: 5,
               id: 1,
               leftover: "",
               opts: %DynamicSegment{size: 0, value: ""},
               proto: 6,
               src_ip: <<10, 10, 10, 5>>,
               srvc_type: 8,
               tot_len: 100,
               ttl: 32,
               vsn: 4
             }

      assert data |> IpDatagram.encode() == bits
    end
  end

  describe "__bit_block__" do
    test "segments" do
      assert segments = DynamicBlock.__bit_block__(:segments)
      assert length(segments) == 8
      assert segments == [:header, :a, :b, :c, :tail, :d, :e, :extra]
    end

    test "segment_info" do
      refute DynamicBlock.__bit_block__(:segment_info, :invalid)

      assert DynamicBlock.__bit_block__(:segment_info, :header) == %{
               default: nil,
               endian: :big,
               name: :header,
               sign: :unsigned,
               size: 4,
               type: :binary
             }
    end
  end
end


mix.exs
defmodule Bitcraft.MixProject do
  use Mix.Project

  @source_url "https://github.com/alboratech/bitcraft"
  @version "0.1.0"

  def project do
    [
      app: :bitcraft,
      version: @version,
      elixir: "~> 1.9",
      elixirc_paths: elixirc_paths(Mix.env()),
      aliases: aliases(),
      deps: deps(),

      # Docs
      name: "Bitcraft",
      docs: docs(),

      # Testing
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        check: :test,
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],

      # Dialyzer
      dialyzer: dialyzer(),

      # Hex
      package: package(),
      description: """
      A toolkit for encoding/decoding bit strings and DSL for binary protocols bit blocks.
      """
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  def application, do: []

  defp deps do
    [
      # Test & Code Analysis
      {:excoveralls, "~> 0.13", only: :test},
      {:credo, "~> 1.5", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.0", only: [:dev, :test], runtime: false},
      {:sobelow, "~> 0.10", only: [:dev, :test], runtime: false},

      # Docs
      {:ex_doc, "~> 0.23", only: :dev, runtime: false}
    ]
  end

  defp aliases do
    [
      check: [
        "compile --warnings-as-errors",
        "format --check-formatted",
        "credo --strict",
        "coveralls.html",
        "sobelow --exit --skip",
        "dialyzer --format short"
      ]
    ]
  end

  defp package do
    [
      name: :bitcraft,
      maintainers: [
        "Carlos Bolanos (GH: cabol)",
        "Albora Technologies"
      ],
      licenses: ["MIT"],
      links: %{"GitHub" => @source_url}
    ]
  end

  defp docs do
    [
      main: "Bitcraft",
      source_ref: "v#{@version}",
      canonical: "http://hexdocs.pm/bitcraft",
      source_url: @source_url
    ]
  end

  defp dialyzer do
    [
      plt_file: {:no_warn, "priv/plts/" <> plt_file_name()},
      flags: [
        :unmatched_returns,
        :error_handling,
        :race_conditions,
        :no_opaque,
        :unknown,
        :no_return
      ]
    ]
  end

  defp plt_file_name do
    "dialyzer-#{Mix.env()}-#{System.otp_release()}-#{System.version()}.plt"
  end
end


lib/bitcraft/helpers.ex
defmodule Bitcraft.Helpers do
  @moduledoc """
  Module for building extra helper functions.
  """

  alias __MODULE__

  @doc false
  defmacro __using__(_opts) do
    quote do
      unquote(Helpers.build_segment_decoder())
    end
  end

  @doc """
  Helper function used internally for building `decode_segment/5` function.
  """
  @spec build_segment_decoder :: term
  def build_segment_decoder do
    sign_opts = [:signed, :unsigned]
    endian_opts = [:big, :little]

    int_exprs = integer_exprs(sign_opts, endian_opts)
    float_exprs = float_exprs(endian_opts)
    bin_exprs = bin_exprs()
    unicode_exprs = unicode_exprs(endian_opts)
    array_exprs = array_exprs()

    {dec_exprs, enc_exprs} =
      [int_exprs, float_exprs, array_exprs, bin_exprs, unicode_exprs]
      |> List.flatten()
      |> Enum.unzip()

    for expr <- dec_exprs ++ enc_exprs do
      expr = Code.string_to_quoted!(expr)

      quote do
        unquote(expr)
      end
    end
  end

  ## Internal Helpers

  defp integer_exprs(sign_opts, endian_opts) do
    for sign <- sign_opts, endian <- endian_opts do
      dec = """
      def decode_segment(bits, size, :integer, :#{sign}, :#{endian}) do
        <<segment::#{sign}-#{endian}-integer-size(size), rest::bits>> = bits
        {segment, rest}
      end
      """

      enc = """
      def encode_segment(var, size, :integer, :#{sign}, :#{endian}) do
        <<var::#{sign}-#{endian}-integer-size(size)>>
      end
      """

      {dec, enc}
    end
  end

  defp float_exprs(endian_opts) do
    for endian <- endian_opts do
      dec = """
      def decode_segment(bits, size, :float, _, :#{endian}) do
        <<segment::#{endian}-float-size(size), rest::bits>> = bits
        {segment, rest}
      end
      """

      enc = """
      def encode_segment(var, size, :float, _, :#{endian}) do
        <<var::#{endian}-float-size(size)>>
      end
      """

      {dec, enc}
    end
  end

  defp bin_exprs do
    for type <- [:bitstring, :bits, :binary, :bytes] do
      dec = """
      def decode_segment(bits, size, :#{type}, _, _) do
        <<segment::#{type}-size(size), rest::bits>> = bits
        {segment, rest}
      end
      """

      enc = """
      def encode_segment(var, _, :#{type}, _, _) do
        <<var::#{type}>>
      end
      """

      {dec, enc}
    end
  end

  defp unicode_exprs(endian_opts) do
    for type <- [:utf8, :utf16, :utf32], endian <- endian_opts do
      dec =
        if type == :utf8 do
          """
          def decode_segment(bits, _, :utf8, _, :#{endian}) do
            if is_integer(bits) do
              <<segment::#{type}-#{endian}, rest::bits>> = bits
              {segment, rest}
            else
              {:unicode.characters_to_binary(bits, :utf8), ""}
            end
          end
          """
        else
          """
          def decode_segment(bits, _, :#{type}, _, :#{endian}) do
            if is_integer(bits) do
              <<segment::#{type}-#{endian}, rest::bits>> = bits
              {segment, rest}
            else
              {:unicode.characters_to_binary(bits, {:#{type}, :#{endian}}), ""}
            end
          end
          """
        end

      enc = """
      def encode_segment(var, _, :#{type}, _, :#{endian}) do
        if is_integer(var) do
          <<var::#{type}-#{endian}>>
        else
          for << ch <- var >>, into: <<>>, do: <<ch::#{type}-#{endian}>>
        end
      end
      """

      {dec, enc}
    end
  end

  defp array_exprs do
    dec = """
    def decode_segment(bits, sz, %Bitcraft.BitBlock.Array{type: t, element_size: esz}, sign, endian) do
      {array, rest} =
        Enum.reduce(1..(div(sz, esz)), {[], bits}, fn _, {lst, bin} ->
          {value, bin} = Bitcraft.decode_segment(bin, esz, t, sign, endian)
          {[value | lst], bin}
        end)

      {Enum.reverse(array), rest}
    end
    """

    enc = """
    def encode_segment(array, _, %Bitcraft.BitBlock.Array{type: t, element_size: s}, sign, endian) do
      for elem <- array, into: <<>> do
        encode_segment(elem, s, t, sign, endian)
      end
    end
    """

    {dec, enc}
  end
end


lib/bitcraft.ex
defmodule Bitcraft do
  @moduledoc """
  The following are the main Bitcraft components:

    * `Bitcraft.BitBlock` - This is the main Bitcraft component. It provides
      a DSL that allows to define bit-blocks with their segments (useful for
      building binary protocols) and automatically injects encoding and decoding
      functions for them.
    * `Bitcraft` - This is a helper module that provides utility functions to
      work with bit strings and binaries.

  """

  use Bitwise
  use Bitcraft.Helpers

  # Base data types for binaries
  @type base_type ::
          integer
          | float
          | binary
          | bitstring
          | byte
          | char

  @typedoc "Segment type"
  @type segment_type :: base_type | Bitcraft.BitBlock.Array.t()

  @typedoc "Codable segment type"
  @type codable_segment_type :: base_type | [base_type]

  ## API

  @doc """
  Encodes the given `input` into a bitstring.

  ## Options

    * `:size` - The size in bits for the input to encode. The default
      value depend on the type, for integer is 8, for float is 63, and for
      other data types is `nil`. If the `input` is a list, this option is
      skipped, since it is handled as array and the size will be
      `array_length * element_size`.

    * `:type` - The segment type given by `Bitcraft.segment_type()`.
      Defaults to `:integer`.

    * `:sign` - If the input is an integer, defines if it is `:signed`
      or `:unsigned`. Defaults to `:unsigned`.

    * `:endian` - Applies to `utf32`, `utf16`, `float`, `integer`.
      Defines the endianness, `:big` or `:little`. Defaults to `:big`.

  ## Example

      iex> Bitcraft.encode_segment(15)
      <<15>>

      iex> Bitcraft.encode_segment(255, size: 4)
      <<15::size(4)>>

      iex> Bitcraft.encode_segment(-3.3, size: 64, type: :float)
      <<192, 10, 102, 102, 102, 102, 102, 102>>

      iex> Bitcraft.encode_segment("hello", type: :binary)
      "hello"

      iex> Bitcraft.encode_segment(<<1, 2, 3>>, type: :bits)
      <<1, 2, 3>>

      iex> Bitcraft.encode_segment([1, -2, 3], type: %Bitcraft.BitBlock.Array{
      ...>   type: :integer, element_size: 4},
      ...>   sign: :signed
      ...> )
      <<30, 3::size(4)>>
  """
  @spec encode_segment(codable_segment_type, Keyword.t()) :: bitstring
  def encode_segment(input, opts \\ []) do
    type = Keyword.get(opts, :type, :integer)
    sign = Keyword.get(opts, :sign, :unsigned)
    endian = Keyword.get(opts, :endian, :big)
    size = Keyword.get(opts, :size)

    size =
      cond do
        is_nil(size) and is_integer(input) -> 8
        is_nil(size) and is_float(input) -> 64
        true -> size
      end

    encode_segment(input, size, type, sign, endian)
  end

  @doc """
  Returns a tuple `{decoded value, leftover}` where the first element is the
  decoded value from the given `input` (according to the given `otps` too)
  and the second element is the leftover.

  ## Options

    * `:size` - The size in bits to decode. Defaults to `byte_size(input) * 8`.
      If the type is `Bitcraft.BitBlock.Array.()`, the size should match with
      `array_length * element_size`.

    * `:type` - The segment type given by `Bitcraft.segment_type()`.
      Defaults to `:integer`.

    * `:sign` - If the input is an integer, defines if it is `:signed`
      or `:unsigned`. Defaults to `:unsigned`.

    * `:endian` - Applies to `utf32`, `utf16`, `float`, `integer`.
      Defines the endianness, `:big` or `:little`. Defaults to `:big`.

  ## Example

      iex> 3
      ...> |> Bitcraft.encode_segment(size: 4)
      ...> |> Bitcraft.decode_segment(size: 4)
      {3, ""}

      iex> -3.3
      ...> |> Bitcraft.encode_segment(size: 64, type: :float, sign: :signed)
      ...> |> Bitcraft.decode_segment(size: 64, type: :float, sign: :signed)
      {-3.3, ""}

      iex> "test"
      ...> |> Bitcraft.encode_segment(type: :binary)
      ...> |> Bitcraft.decode_segment(size: 4, type: :binary)
      {"test", ""}

      iex> <<1, 2, 3, 4>>
      ...> |> Bitcraft.encode_segment(type: :bits)
      ...> |> Bitcraft.decode_segment(size: 32, type: :bits)
      {<<1, 2, 3, 4>>, ""}

      iex> alias Bitcraft.BitBlock.Array
      iex> [1, 2]
      ...> |> Bitcraft.encode_segment(type: %Array{})
      ...> |> Bitcraft.decode_segment(size: 16, type: %Array{})
      {[1, 2], ""}
      iex> [3.3, -7.7, 9.9]
      ...> |> Bitcraft.encode_segment(
      ...>   type: %Array{type: :float, element_size: 64},
      ...>   sign: :signed
      ...> )
      ...> |> Bitcraft.decode_segment(
      ...>   size: 192,
      ...>   type: %Array{type: :float, element_size: 64},
      ...>   sign: :signed
      ...> )
      {[3.3, -7.7, 9.9], ""}
  """
  @spec decode_segment(bitstring, Keyword.t()) :: {codable_segment_type, bitstring}
  def decode_segment(input, opts \\ []) do
    type = Keyword.get(opts, :type, :integer)
    sign = Keyword.get(opts, :sign, :unsigned)
    endian = Keyword.get(opts, :endian, :big)
    size = opts[:size] || byte_size(input) * 8

    decode_segment(input, size, type, sign, endian)
  end

  @doc """
  Returns the number of `1`s in binary representation of the given `integer`.

  ## Example

      iex> Bitcraft.count_ones(15)
      4

      iex> Bitcraft.count_ones(255)
      8
  """
  @spec count_ones(integer) :: integer
  def count_ones(integer) when is_integer(integer) do
    count_ones(integer, 0)
  end

  defp count_ones(0, count), do: count

  defp count_ones(integer, count) do
    count_ones(integer &&& integer - 1, count + 1)
  end
end
